### PR TITLE
143 better validations

### PR DESCRIPTION
**Pull request by EdwardAndress**
https://trello.com/c/J4RP5Rwh/143-better-validations-both-in-public-and-staff-app

I've also done a little refactoring here, in the initial commits, to implement shoulda matchers and break up some long lines.

### CODE REVIEW

File: spec/models/correspondence_spec.rb [hunk 1]
```diff
@@ -19,31 +19,29 @@
...
-      it 'name' do
-        correspondence.name = nil
-        expect(correspondence).not_to be_valid
-      end
-
-      it 'email' do
-        correspondence.email = nil
-        expect(correspondence).not_to be_valid
-      end
+    context 'mandatory' do
```
Comment (misaka) [inline spec/models/correspondence_spec.rb]: I'm not sure I understand the use of the `contexts` here. Contexts should describe a situation and (usually) should be reflected in changes to the setup of fixtures, etc. Here they seem to be a way of grouping tests that don't really need it, especially with the `should` syntax you've now added.

File: spec/rails_helper.rb [hunk 2]
```diff
@@ -59,3 +59,10 @@
   # arbitrary gems may also be filtered via:
   # config.filter_gems_from_backtrace("gem name")
 end
+
+Shoulda::Matchers.configure do |config|
```
Comment (misaka) [inline spec/rails_helper.rb]: Is `Shoulda` here to support the one-liner syntax? Is this not built-into `rspec` using it's [one-liner expectations](https://github.com/rspec/rspec-expectations/blob/master/Should.md#one-liners)?

Comment (aliuk2012) [inline spec/rails_helper.rb]: Its part of the shoulda-matchers gem configuration. I'll also be using it for the feedback branch.  In short there are some methods that let you write things like this 

```
  it { should validate_inclusion_of(:type).in_array(%w( feedback bug_report )) }
```
validate_inclusion_of is not part of rspec

Comment (misaka) [inline spec/rails_helper.rb]: @aliuk2012 What about the page I linked-to that seems to suggest that they're supported?

File: spec/models/correspondence_spec.rb (hunk 1 above)
Comment (EdwardAndress) [inline spec/models/correspondence_spec.rb]: I see that it was a little misleading.  The intention was to break up the tests into groups for readability.  This was OK, in my opinion, using the original syntax but is no longer needed now that most are one-liners.

File: spec/rails_helper.rb (hunk 2 above)
Comment (EdwardAndress) [inline spec/rails_helper.rb]: When I remove this code from rails_helper, a bunch of tests fail

File: spec/controllers/correspondence_controller_spec.rb [hunk 3]
```diff
@@ -17,26 +22,25 @@
...
+      it { should render_template(:confirmation)}
 
       it 'sends #perorm_later to the EmailCorrespondenceJob' do
         expect(EmailCorrespondenceJob).to receive(:perform_later)
         post :create, params: params        
       end
 
       it 'and a job is enqueued' do
-        expect { post :create, params: params }.to have_enqueued_job(EmailCorrespondenceJob)
+        expect {
```
Comment (EdwardAndress) [inline spec/controllers/correspondence_controller_spec.rb]: I wasn't sure how to tag on the #to method call, in a nice way, using `do` and `end` . I don't like `end.to have_enqueued` ... so I've broken the line after the block instead

File: spec/rails_helper.rb (hunk 2 above)
Comment (EdwardAndress) [inline spec/rails_helper.rb]: It's a good point @misaka .  I can't, yet, explain why the one-liner syntax described on the linked page seems to fail.  I've tried a few things and read around but always get the same error/s:
- undefined method `validate_presence_of' (also for other matchers used in one liners)

Comment (EdwardAndress) [inline spec/rails_helper.rb]: The docs would imply that all I need do is declare the subject (which I have done).  All searches lead to the shoulda-matchers.  None lead to any explanation as to why the RSpec one-liner syntax fails (which does make me think I am doing something very wrong!).

Comment (EdwardAndress) [inline spec/rails_helper.rb]: I am curious to know the answer but think we { should go_with_what_works }.for_now

File: spec/models/correspondence_spec.rb [hunk 4]
```diff
@@ -2,12 +2,10 @@
 
 Rails.describe Correspondence, type: :model do
```
Comment (misaka) [inline spec/models/correspondence_spec.rb]: Shouldn't this be `RSpec.describe`?

File: spec/models/correspondence_spec.rb [hunk 5]
```diff
@@ -19,32 +17,24 @@
...
+    end
 
-      it 'message' do
-        correspondence.message = nil
-        expect(correspondence).not_to be_valid
-      end
+    it do
+      should validate_inclusion_of(:type).
+        in_array(Settings.correspondence_types)
+    end
```
Comment (misaka) [inline spec/models/correspondence_spec.rb]: Trailing whitespace here.

File: spec/rails_helper.rb (hunk 2 above)
Comment (EdwardAndress) [inline spec/rails_helper.rb]: Indeed!  I also found that.  Let's make a note and send them a sweet PR.

File: spec/models/correspondence_spec.rb (hunk 4 above)
Comment (EdwardAndress) [inline spec/models/correspondence_spec.rb]: Yes.  I thought I had changed it.  Well spotted.

**Pull request merged**